File: common/src/main/java/com/ultreon/mods/advanceddebug/client/registry/FormatterRegistry.java
package com.ultreon.mods.advanceddebug.client.registry;

import com.ultreon.libs.collections.v0.maps.OrderedHashMap;
import com.ultreon.mods.advanceddebug.api.client.menu.Formatter;
import com.ultreon.mods.advanceddebug.api.client.registry.IFormatterRegistry;

import javax.annotation.Nullable;
import java.util.Map;

public class FormatterRegistry implements IFormatterRegistry {
    private static final Map<String, Formatter<?>> FORMATTERS = new OrderedHashMap<>();
    private static final FormatterRegistry INSTANCE = new FormatterRegistry();

    private FormatterRegistry() {

    }

    public static FormatterRegistry get() {
        return INSTANCE;
    }

    @Override
    public <T> Formatter<T> register(Formatter<T> formatter) {
        System.out.println("formatter = " + formatter);
        Class<?> clazz = formatter.clazz();
        FORMATTERS.put(clazz.getName(), formatter);
        return formatter;
    }

    public void dump() {
        System.out.println("-=====- DEBUG FORMATTER REGISTRY DUMP -=====-");
        for (Map.Entry<String, Formatter<?>> entry : FORMATTERS.entrySet()) {
            System.out.println(entry.getKey());
            System.out.println(entry.getValue().registryName());
        }
        System.out.println("_______ DEBUG FORMATTER REGISTRY DUMP _______");
    }

    @Override
    @Nullable
    public Formatter<?> identify(Class<?> aClass) {
        for (Class<?> clazz = aClass; clazz != null; clazz = clazz.getSuperclass()) {
            if (FORMATTERS.containsKey(clazz.getName())) {
                return FORMATTERS.get(clazz.getName());
            }
            for (Class<?> inter : clazz.getInterfaces()) {
                Formatter<?> identify = identify(inter);
                if (identify != null) return identify;
            }
        }
        return null;
    }
}

File: common/src/main/kotlin/com/ultreon/mods/advanceddebug/client/registry/FormatterRegistry.kt
package com.ultreon.mods.advanceddebug.client.registry

import com.ultreon.mods.advanceddebug.api.client.menu.Formatter
import com.ultreon.mods.advanceddebug.api.client.registry.IFormatterRegistry
import kotlin.reflect.KClass

class FormatterRegistry private constructor() : IFormatterRegistry {
    override fun <T : Any> register(formatter: Formatter<T>): Formatter<T> {
        val clazz: Class<*> = formatter.javaClass()
        FORMATTERS[clazz.name] = formatter
        return formatter
    }

    fun dump() {
        println("-=====- DEBUG FORMATTER REGISTRY DUMP -=====-")
        for ((key, value) in FORMATTERS) {
            println(key)
            println(value!!.registryName())
        }
        println("_______ DEBUG FORMATTER REGISTRY DUMP _______")
    }

    override fun identify(`class`: Class<*>): Formatter<*>? {
        var clazz: Class<*>? = `class`
        while (true) {
            if (FORMATTERS.containsKey(clazz!!.name)) {
                return FORMATTERS[clazz.name]
            }
            for (inter in clazz.interfaces) {
                val identify = identify(inter)
                if (identify != null) return identify
            }
            if (clazz.superclass != null) {
                clazz = clazz.superclass
            } else {
                return null
            }
        }
    }

    override fun identify(kClass: KClass<*>): Formatter<*>? {
        return identify(kClass.java)
    }

    companion object {
        private val FORMATTERS: MutableMap<String, Formatter<*>?> = HashMap()
        private val INSTANCE = FormatterRegistry()
        fun get(): FormatterRegistry {
            return INSTANCE
        }
    }
}